---
title: "XengKPreProcess.Rmd"
author: "Carly Muletz Wolz"
date: "2/15/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

### Read in data and remove singletons
```{r}

##    Grayfer Lab Xenopus Mast cell enrichment + Bd Study 2022 PRE-process Data Analysis   ###


## Load packages
library(phyloseq)
library(ape)
library(vegan)
library(ggplot2)

##Set working directory to bring in files##
#setwd("/Users/lkgentry/OneDrive - Smithsonian Institution/GrayferLabExperiments/KelseyAnalysisNov2022/FinalRFilesCombinedRun/")

setwd("/Users/Carly/OneDrive - Smithsonian Institution/NSF-BBSRC2022/GrayferLabExperiments/KelseyAnalysisNov2022/FinalRFilesCombinedRun/")

# this is site by species matrix, need row.names = 1 to have phyloseq read it
# there was an issue with original featureTab. Left over _ in sample names and - between controls instead of _
featureTab <- read.csv("XENGK2022EXP2_feature_table_run2.csv", header = T, row.names = 1)

# make compatible for phyloseq format
featureTab = otu_table(featureTab, taxa_are_rows = TRUE)
## 6129 taxa by 66 samples
dim(featureTab)

# Read taxonomy info in, make matrix and compatible for phyloseq
taxonomy <- tax_table(as.matrix(read.csv("XENGK2022EXP2_taxonomy_run2.csv", row.names = 1)))


meta_data <- sample_data(read.csv("XENGK2022EXP2_META.csv", header = T, row.names = 1))

## SampleID is now row.names, still useful to have SampleID in metadata so add in

meta_data$SampleID <- row.names(meta_data)


#did not run this
#Read in sequence data, may need if you want to look at or subset the DNA sequences at some point
#library(Biostrings)
#seqs <- readDNAStringSet("XENGK2022EXP2_feature_DNAsequences_run2.fasta")

# You can also add a phylogenetic tree here, if you have one
# library(ape)
# tree = read.tree("FinalRFiles/exported-tree/SalAMPtree.nwk")

# Merge it all together

xk <- merge_phyloseq(featureTab, taxonomy, meta_data) #seqs, #tree)

xk



## Check all samples are coming in

## 1645 taxa by 66 samples
dim(featureTab)
sample_names(featureTab)

sample_names(xk)

sum(sample_sums(xk))
#883797

sort(sample_sums(xk))

## Filter singletons (only occur on 1 individual), seems to be a lot of ASVs with little information
## This says needs to occur at least 1 time on at least 2 individual

xk2 <- filter_taxa(xk, function (x) {sum(x > 0) >1}, prune=TRUE)

## Note 5723 taxa and now 1791. I always remove singletons to individuals (note some people call singletons just 1 sequence) as I believe many are spurious. We can check the standards to verify this also.
xk
xk2

## Lost a good bit of taxa, but not many sequences

sum(sample_sums(xk))
sum(sample_sums(xk2))

#now 829411 for xk2 sum

sort(sample_sums(xk2))


## REMOVE chloroplast and eukaryotic sequences


get_taxa_unique(xk2, "Kingdom")

## The NA may be host 16S rRNA, could blast if of interest
Euk <- subset_taxa(xk2, Kingdom %in% c("Eukaryota", NA))

tax_table(Euk)

xk3 = subset_taxa(xk2, Kingdom %in% c("Bacteria", "Archaea"))

xk3 <- subset_taxa(xk3, Class != "Chloroplast")

xk3
```


### Let's look at the positive standards first, then remove for contaminant filtering

### We recover all the bacterial taxa that should be there and in relatively similar relative abundances to what is expected, but there are extra that shouldn't be there

#### See ZymoStandards-Taxonomy-Compare for more info
```{r}
MicroStClean <- subset_samples(xk3, Sample_or_Control %in% c("Standard Extraction", "Standard PCR"))

## Are there ASVs that are lingering?
sum(taxa_sums(MicroStClean) == 0)

## getting rid of ASvs that are not in samples of interest
MicroStClean = filter_taxa(MicroStClean, function(x) sum(x) !=0, TRUE)

MicroStClean <- transform_sample_counts(MicroStClean, function(x) x/sum(x))

otu_table(MicroStClean)
dfo <- as.data.frame(otu_table(MicroStClean))

## column 6 is genus, or change to the column of genus
tax_table(MicroStClean)[,6]

dft <- as.data.frame(tax_table(MicroStClean)[,6])

dfpc <- cbind(dfo, dft)

dfpc[order(dfpc$Genus), ]


#Check positive controls have correct bacteria present. Can look in feature table two and taxanomy files. Will also tell you below. Take into account there can be different names for same bacteria species due to re-classification
### Genus	Truth, same for both extraction and PCR controls
#Bacillus	0.18
#Enterococcus	0.101
#Escherichia/Shigella	0.15
#Lactobacillus	0.12
#Listeria	0.141
#Pseudomonas	0.062
#Salmonella	0.10
#Staphylococcus	0.15


#Controls look really good
## Let's see if decontam removes them

## can write a file too and make plot

site_species <-as(otu_table(MicroStClean), "matrix")
taxonomy <- as(tax_table(MicroStClean), "matrix")
write.csv(cbind(site_species, taxonomy), "ZymoStandards.csv")


st <- psmelt(MicroStClean) # create dataframe from phyloseq object

st <- subset(st, Abundance > 0)
library(ggplot2)
plot.rel.ab2 <- ggplot(data=st, aes(x=Sample, y=Abundance, fill=Genus))
plot.rel.ab2 + geom_bar(aes(), stat="identity", position="stack")  +
  ylab("Relative abundance (% of total sequences)") +theme_classic() + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + theme_bw()  +
  theme_classic()+ labs(x = "")+ theme(text = element_text(size = 20))  





### Let's come back and look at these after filtering contaminants. Copied below when you need to look at it again


## Go back to before we removed singletons, interested if singletons are showing up in standards...
MicroStPrePre <- subset_samples(xk, Sample_or_Control %in% c("Standard Extraction", "Standard PCR"))

MicroStPrePre <- filter_taxa(MicroStPrePre, function(x) sum(x) !=0, TRUE)

otu_table(MicroStPrePre)

tax_table(MicroStPrePre)

### Hmmmm...wow, good to check. Important to remove singletons!

```






### Contaminant filter
#### See more: https://benjjneb.github.io/decontam/vignettes/decontam_intro.html

```{r}
####----Decontam----####
citation ("decontam")


library(decontam)



## Says in paper that combined method provides the best bimodal distribution 
## suggesting that is will be the more robust classification when both data types
## are present, which we have


## NOTE: you need to put in your DNA quant readings that you used for pooling your library into the metadata file and call it quant_reading. Look a the tutorial indicated above
## ALSO later you need a column that says Sample_or_Control

## Keep positive controls in right now, will relook at and assess

sample_data(xk3)$is.neg <- sample_data(xk3)$Sample_or_Control == "Control Sample"

## Using combined based on paper recommendation
contamdf.combined <- isContaminant(xk3, method="combined", neg="is.neg", conc = "quant_reading")
#removed 1 sample with 0 total counts or frequency

## If error, make sure all have quant readings. For too low qubit readings - I put 0.05. In this data set, all samples have higher quant readings
sample_data(xk3)$quant_reading
table(contamdf.combined$contaminant)

hist(contamdf.combined$p)


ps.pa <- transform_sample_counts(xk3, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$Sample_or_Control == "Control Sample", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$Sample_or_Control == "True Sample", ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.combined$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

## Looks like most in neg control are exclusive to neg control (= FALSE), then a couple in a lot of samples, but that's ok. Do check if you lose a lot of sequences after this if there is an issue with what decontam is removing.

## row numbers of some of the contaminants (NOT ASV numbers)
head(which(contamdf.combined$contaminant))

## says 666, 1154, 1176, 1241, 1249, 1354. So let's plot some that are contaminants and some that aren't

plot_frequency(xk3, taxa_names(xk3)[c(666,1154,1176,1241,1249,1354,22,23,24)], conc="quant_reading") + 
  xlab("DNA Concentration")

##come back to this
#Error in data.frame(..., check.names = FALSE) : 
#  arguments imply differing number of rows: 65, 66
#In addition: Warning message:
#In plot_frequency(xk3, taxa_names(xk3)[c(666, 1154, 1176, 1241,  :
#  Removed 1 samples with zero total counts (or frequency).

####----Delete Contaminants----####
#Need to delete ASVs found as contaminants by either method

xk4 <- prune_taxa(!contamdf.combined$contaminant, xk3)

sum(sample_sums(xk3))
sum(sample_sums(xk4))
sort(sample_sums(xk4))


xk4 <- subset_samples(xk4, Sample_or_Control != "Control Sample")

sort(sample_sums(xk4))

## Let's look at positive control again, see if contaminant filtering helped


MicroStClean2 <- subset_samples(xk4, Sample_or_Control %in% c("Standard Extraction", "Standard PCR"))

## getting rid of ASvs that are not in samples of interest
MicroStClean2 = filter_taxa(MicroStClean2, function(x) sum(x) !=0, TRUE)

otu_table(MicroStClean2)

otu_table(MicroStClean)

tax_table(MicroStClean2)
## Still same ones.### Genus	Truth, same for both extraction and PCR controls
#forgot where these numbers come from but believe its matching ASVs to taxa
#Bacillus	0.199
#Enterococcus	0.1
#Escherichia/Shigella	0.13
#Lactobacillus	0.13
#Listeria	0.12
#Pseudomonas	0.06
#Salmonella	0.11
#Staphylococcus	0.15



## can write to file too and make plot

site_species <-as(otu_table(MicroStClean2), "matrix")
taxonomy <- as(tax_table(MicroStClean2), "matrix")
write.csv(cbind(site_species, taxonomy), "ZymoStandards.csv")

## Need to call new name next time
write.csv(as(otu_table(xk4), "matrix"), "XENGK2022EXP2_feature_table_run2.csv")

## Positive controls looked really good. D21Vbd8 oddly has positive control sequences, concerned about contamination.
## Can see this in XENGK2022EXP2_feature_table2_run2.
## Manually deleting the D21VBd8 sample and then the Positive controls and saving as XENGK2022EXP2_feature_tableFINAL_run2.csv


```

######### ALPHA DIVERSITY Calculate and add to mapping file with seq counts  ############

```{r Estimate richess}

featureTab2 <- read.csv("XENGK2022EXP2_feature_tableFINAL_run2.csv", header = T, row.names = 1)

# make compatible for phyloseq format
featureTab2 = otu_table(featureTab2, taxa_are_rows = TRUE)
## 1645 taxa by 53 samples
dim(featureTab2)

# Merge it all together

xk6 <- merge_phyloseq(featureTab2, taxonomy, meta_data)# seqs) #tree)

xk6
dfxk6 <- as(sample_data(xk6), "data.frame")

t_otu <-t(as(otu_table(xk6), "matrix"))

library(vegan)
AD <- estimateR(t_otu)
AD <- t(as.data.frame(AD))

#need to have both alpha and df having the same column info
# this won't work with estimateR, but is if you use PD estimate for phylogenetic diversity
#PD$SampleID <- row.names(PD)

## Add sequence coverage info
seqs <- as.data.frame(sample_sums(xk6))
seqs$SampleID <- row.names(seqs)

#now merge to get sequence counts and alpha estimates in mapping file
alpha_df <- merge(dfxk6, AD,  by = "row.names")
alpha_df$SampleID <- alpha_df$Row.names
alpha_df <- merge(alpha_df, seqs, by = "SampleID")
row.names(alpha_df) <- alpha_df$Row.names


## No huge sequence count difference, no rarefying
sort(sample_sums(xk6))
sum(sample_sums(xk6))



max(sample_sums(xk6)) #24584
min(sample_sums(xk6)) #1842
max(sample_sums(xk6))/min(sample_sums(xk6))

#13.34x is large difference. Most likely due to lower quality sequence run but much better than before (went down from 48.5x). Will need to rarify later. You want this number to be less than 5x

## NOTE: once you make this file you can read it at the beginning and you don't need to do these calculations again

## NOTE: I also like phylogenetic diversity to estimate, but you need a bacterial tree for that and in this example, I did not generate one. I have code if you would like. I make the tree in QIIME. I have not found a way to do it in R that works. 

## This now is your clean file. All controls are removed 

write.csv(alpha_df, "XENGK2022EXP2_meta_FINAL_run2.csv", row.names = F)


## NEXT TIME MOVE ANTI-BD bac calculations here

```






